README.md
# eimpute: Efficiently IMPUTE Large Scale Incomplete Matrix

Introdution
----------
Matrix completion is a procedure for imputing the missing elements in matrices by using the information of observed elements. This procedure can be visualized as:

![](./vignettes/matrixcom.jpg)

Matrix completion has attracted a lot of attention, it is widely applied in:
- tabular data imputation: recover the missing elements in data table;
- recommend system: estimate users' potantial preference for items pending purchased;
- image inpainting: inpaint the missing elements in digit images.


Software
----------
A computationally efficient R package, **eimpute** is developed for matrix completion.

### Installation
Install the stable version from CRAN:        
```R
install.packages("eimpute")
```

### Advantage
In **eimpute**, matrix completion problem is solved by iteratively performing low-rank approximation and data calibration, which enjoy two admirable advantages:
- unbiased low-rank approximation for incomplete matrix
- less time consumption via truncated SVD 
Moreover, **eimpute** also supports flexible data standardization.

Compare **eimpute** and **softimpute** in systhesis datasets $X_{m \times m}$ with $p$ proportion missing observations:

- $m$ is chosen as 1000, 2000, 3000, 4000
- $p$ is chosen as 0.1, 0.5, 0.9.

![](./vignettes/time3.png)
![](./vignettes/error3.png)

In high dimension case, als method in **softimpute** is a little faster than **eimpute** in low proportion of missing observations, as the proportion of missing observations increase, rsvd method in **eimpute** have a better performance than **softimpute** in time cost and test error. Compare with two method in **eimpute*, rsvd method is better than tsvd in time cost.

References
----------
- Rahul Mazumder, Trevor Hastie and Rob Tibshirani (2010) Spectra Regularization Algorithms for Learning Large Incomplete Matrices, Journal of Machine Learning Research 11 (2010) 2287-2322

- Nathan Halko, Per-Gunnar Martinsson, Joel A. Tropp (2011) Finding Structure with Randomness: Probabilistic Algorithms for Constructing Approximate Matrix Decompositions, SIAM Review Vol. 53, num. 2, pp. 217-288

Bug report
----------
Send an email to Zhe Gao at gaozh8@mail2.sysu.edu.cn


NEWS.md
# eimpute 0.2.1

* Add a parameter of initialization matrix to eimpute and r.search function.

# eimpute 0.1.1

* Add a rank search function with information criterion rule and cross validation.
* Use Rcpp to rewrite biscale function.

# eimpute 0.1.0

* Initial CRAN version.


R/eimpute.R
#' @title Efficiently impute missing values for a large scale matrix
#'
#' @description Fit a low-rank matrix approximation to a matrix with missing values.
#' The algorithm iterates like EM: filling the missing values with the current guess,
#' and then approximating the complete matrix via truncated SVD.
#'
#' @aliases eimpute
#'
#' @param x an \eqn{m} by \eqn{n} matrix with \code{NA}s.
#' @param r the rank of low-rank matrix for approximating \code{x}
#' @param svd.method a character string indicating the truncated SVD method.
#' If \code{svd.method = "rsvd"}, a randomized SVD is used,
#' else if \code{svd.method = "tsvd"}, standard truncated SVD is used.
#' Any unambiguous substring can be given. Default \code{svd.method = "tsvd"}.
#' @param noise.var the variance of noise.
#' @param init if init = FALSE(the default), the missing entries will initialize with mean.
#' @param init.mat the initialization matrix.
#' @param thresh convergence threshold, measured as the relative change in the Frobenius norm between two successive estimates.
#' @param maxit	 maximal number of iterations.
#' @param override logical value indicating whether the observed elements in \code{x} should be overwritten by its low-rank approximation.
#' @param control a list of parameters that control details of standard procedure, See \link{biscale.control}.
#' @param ... arguments to be used to form the default control argument if it is not supplied directly.
#'
#'
#' @return A list containing the following components
#' \item{\code{x.imp}}{the matrix after completion.}
#' \item{\code{rmse}}{the relative mean square error of matrix completion, i.e., training error.}
#' \item{\code{iter.count}}{the number of iterations.}
#'
#' @references Rahul Mazumder, Trevor Hastie and Rob Tibshirani (2010) Spectral Regularization Algorithms for Learning Large Incomplete Matrices, Journal of Machine Learning Research 11, 2287-2322
#' @references Nathan Halko, Per-Gunnar Martinsson, Joel A. Tropp (2011) Finding Structure with Randomness: Probabilistic Algorithms for Constructing Approximate Matrix Decompositions, Siam Review Vol. 53, num. 2, pp. 217-288
#'
#' @rdname eimpute
#' @useDynLib eimpute, .registration = TRUE
#' @importFrom Rcpp sourceCpp
#' @export
#' @examples
#' ################# Quick Start #################
#' m <- 100
#' n <- 100
#' r <- 10
#' x_na <- incomplete.generator(m, n, r)
#' head(x_na[, 1:6])
#' x_impute <- eimpute(x_na, r)
#' head(x_impute[["x.imp"]][, 1:6])
#' x_impute[["rmse"]]
eimpute <- function(x, r, svd.method = c("tsvd", "rsvd"), noise.var = 0,
                    thresh = 1e-05, maxit = 100, init = FALSE, init.mat = 0, override = FALSE, control = list(...), ...){
  m <- nrow(x)
  n <- ncol(x)

  x_sd <- biscale(x, control = control)

  x_train <- x_sd[[1]]
  # x_train <- x
  ind_ob <- which(!is.na(x_train), arr.ind = TRUE)
  x_ob <- x_train[ind_ob]
  ind <- ind_ob - 1

  svdm <- match.arg(svd.method)
  type <- ifelse(svdm == "tsvd", 1, 2)

  ind_miss <- which(is.na(x_train), arr.ind = TRUE)



  if(init == TRUE){
    ind_miss <- cbind(ind_miss, as.vector(init.mat[ind_miss]))
  }else{
    ind_miss <- cbind(ind_miss, rnorm(nrow(ind_miss), 0, noise.var))
  }
  ind_miss[,1:2] <- ind_miss[,1:2] - 1

  Z_temp <- kkt_fix(ind, ind_miss, x_ob, m, n, r, maxit, thresh, type, init)
  Z.fit <- Z_temp[[1]] * (x_sd[[4]] %*% t(x_sd[[5]])) + matrix(rep(x_sd[[2]], n), nrow = m) + t(matrix(rep(x_sd[[3]], m), nrow = n))


  if (!override) {
    Z.fit[ind_ob] <- x_ob
  }

  list(x.imp = Z.fit, rmse = Z_temp[[3]], iter.count = Z_temp[[2]])
}


